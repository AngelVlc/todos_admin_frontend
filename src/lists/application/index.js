export * from "./GetListByIdUseCase";
export * from "./DeleteListByIdUseCase";
export * from "./GetListsUseCase";
export * from "./CreateListUseCase";
export * from "./UpdateListUseCase";
export * from "./MoveListItemUseCase";
export * from "./IndexAllListsUseCase";
export * from "./GetSearchSecureKeyUseCase";
