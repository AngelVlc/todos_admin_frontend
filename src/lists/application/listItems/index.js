export * from "./DeleteListItemByIdUseCase";
export * from "./GetListItemByIdUseCase";
export * from "./GetListItemsUseCase";
export * from "./CreateListItemUseCase";
export * from "./UpdateListItemUseCase";
