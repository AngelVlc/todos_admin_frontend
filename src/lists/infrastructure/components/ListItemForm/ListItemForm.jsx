import React, { useRef, forwardRef, useImperativeHandle } from "react";
import { Formik, Form, Field, ErrorMessage } from "formik";
import { ListItem } from "../../../domain";
import * as Yup from "yup";

export const ListItemForm = forwardRef((props, ref) => {
  const formRef = useRef();

  useImperativeHandle(ref, () => ({
    submitForm: async () => {
      await formRef.current.submitForm();
    },
    setValues: (values) => {
      formRef.current.setValues(values, false);
    },
  }));

  return (
    <Formik
      innerRef={formRef}
      initialValues={ListItem.createEmpty(-1)}
      enableReinitialize={true}
      validationSchema={Yup.object({
        title: Yup.string().required("Required"),
      })}
      onSubmit={props.onSubmit}
    >
      <Form>
        <div className="field">
          <label className="label" htmlFor="title">
            Title
          </label>
          <div className="control">
            <Field
              name="title"
              as="input"
              className="input"
              data-testid="title"
              autoFocus={true}
            />
          </div>
          <p className="help is-danger" data-testid="titleErrors">
            <ErrorMessage name="title" />
          </p>
        </div>
        <div className="field">
          <label className="label" htmlFor="description">
            Description
          </label>
          <div className="control">
            <Field
              name="description"
              as="textarea"
              className="textarea"
              data-testid="description"
            />
          </div>
          <p className="help is-danger" data-testid="descriptionErrors">
            <ErrorMessage name="description" />
          </p>
        </div>
      </Form>
    </Formik>
  );
});
