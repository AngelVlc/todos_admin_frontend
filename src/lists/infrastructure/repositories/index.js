export * from "./ListsRepository";
export * from "./ListItemsRepository";
