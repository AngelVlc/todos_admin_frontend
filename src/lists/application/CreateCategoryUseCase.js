import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class CreateCategoryUseCase extends BaseUseCase {
  _categoriesRepository;

  constructor({ categoriesRepository }) {
    super();

    this._categoriesRepository = categoriesRepository;
  }

  static create() {
    const categoriesRepository = new CategoriesRepository();

    return new CreateCategoryUseCase({ categoriesRepository });
  }

  async execute(category) {
   return await this._categoriesRepository.create(category);
  }
}
