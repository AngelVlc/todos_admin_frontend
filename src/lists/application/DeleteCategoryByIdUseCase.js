import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { CategoriesRepository } from "../infrastructure/repositories";

export class DeleteCategoryByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new DeleteCategoryByIdUseCase({ repository });
  }

  async execute(categoryId) {
    return await this._repository.deleteById(categoryId);
  }
}
