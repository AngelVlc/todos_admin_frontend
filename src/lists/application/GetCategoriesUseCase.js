import { BaseUseCase } from "../../shared/domain/BaseUseCase";
import { Category } from "../domain";
import { CategoriesRepository } from "../infrastructure/repositories";

export class GetCategoriesUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new CategoriesRepository();

    return new GetCategoriesUseCase({ repository });
  }

  async execute() {
    const result = await this._repository.getAll();

    return result.map((item) => {
      return new Category(item);
    });
  }
}
