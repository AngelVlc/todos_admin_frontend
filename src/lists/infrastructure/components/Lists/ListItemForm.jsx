import React, { useEffect, useState, useContext } from "react";
import { useHistory } from "react-router-dom";
import { Formik, Form, Field, ErrorMessage } from "formik";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import * as Yup from "yup";
import { ListItem } from "../../../domain";
import {
  CreateListItemUseCase,
  UpdateListItemUseCase,
} from "../../../application/listItems";

export const ListItemForm = (props) => {
  let history = useHistory();
  const { useCaseFactory } = useContext(AppContext);

  const [pageState, setPageState] = useState(
    new ListItem({ title: "", description: "" })
  );

  useEffect(() => {
      setPageState(new ListItem(props.listItem));
  }, [props]);

  const onSubmit = async (listItem) => {
    let useCase;
    if (props.listItem?.id === undefined) {
      useCase = useCaseFactory.get(CreateListItemUseCase);
    } else {
      useCase = useCaseFactory.get(UpdateListItemUseCase);
    }

    const result = await useCase.execute(listItem);
    if (result) {
      history.push(`/lists/${props.listItem.listId}/edit`);
    }
  };

  return (
    <Formik
      enableReinitialize={true}
      initialValues={pageState}
      validationSchema={Yup.object({
        title: Yup.string().required("Required"),
      })}
      onSubmit={onSubmit}
    >
      <Form>
        <div className="field">
          <label className="label" htmlFor="title">
            Title
          </label>
          <div className="control">
            <Field
              name="title"
              as="input"
              className="input"
              data-testid="title"
              autoFocus
            />
          </div>
          <p className="help is-danger" data-testid="titleErrors">
            <ErrorMessage name="title" />
          </p>
        </div>

        <div className="field">
          <label className="label" htmlFor="description">
            Description
          </label>
          <div className="control">
            <Field
              name="description"
              as="textarea"
              className="textarea"
              data-testid="description"
            />
          </div>
          <p className="help is-danger" data-testid="descriptionErrors">
            <ErrorMessage name="description" />
          </p>
        </div>

        <div className="field is-grouped">
          <p className="control">
            <button className="button" data-testid="submit" type="submit">
              {props.listItem?.id ? "SAVE" : "CREATE"}
            </button>
          </p>
          <p className="control">
            <button
              className="button"
              data-testid="cancel"
              type="button"
              onClick={() =>
                history.push(`/lists/${props.listItem.listId}/edit`)
              }
            >
              CANCEL
            </button>
          </p>
          {props.listItem?.id && (
            <p className="control">
              <button
                className="button is-danger"
                data-testid="delete"
                type="button"
                onClick={() =>
                  history.push(
                    `/lists/${props.listItem.listId}/items/${props.listItem.id}/delete`
                  )
                }
              >
                DELETE
              </button>
            </p>
          )}
        </div>
      </Form>
    </Formik>
  );
};
