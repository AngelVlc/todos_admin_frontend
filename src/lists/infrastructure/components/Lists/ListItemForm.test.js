import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { ListItemForm } from "./ListItemForm";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { MemoryRouter, Route } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { CreateListItemUseCase } from "../../../application/listItems";
import { ListItem } from "../../../domain";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const mockedCreateListItemUseCase = {
  execute: jest.fn(),
};

const mockedUpdateListItemUseCase = {
  execute: jest.fn(),
};

const useCaseFactory = {
  get: (useCase) => {
    if (useCase == CreateListItemUseCase) {
      return mockedCreateListItemUseCase;
    }

    return mockedUpdateListItemUseCase;
  },
};

const renderWithContextAndRouterForExistingItem = () => {
  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <MemoryRouter initialEntries={[`/lists/2/items/5/edit`]}>
          <Route path="/lists/:listId/items/:itemId/edit">
            <ListItemForm
              listItem={
                new ListItem({
                  listId: 2,
                  id: 5,
                  title: "item title",
                  description: "item description",
                })
              }
            />
          </Route>
        </MemoryRouter>
      </AppContext.Provider>
    ),
  };
};

const renderWithContextAndRouterForNewItem = () => {
  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <MemoryRouter initialEntries={[`/lists/2/items/new`]}>
          <Route path="/lists/:listId/items/new">
            <ListItemForm listItem={ListItem.createEmpty(2)} />
          </Route>
        </MemoryRouter>
      </AppContext.Provider>
    ),
  };
};

afterEach(cleanup);

it("should match the snapshot for an existing item", async () => {
  let fragment;
  await act(async () => {
    const { asFragment } = renderWithContextAndRouterForExistingItem();
    fragment = asFragment;
  });
  expect(fragment()).toMatchSnapshot();
});

it("should match the snapshot for a new item", async () => {
  const { asFragment } = renderWithContextAndRouterForNewItem();
  expect(asFragment()).toMatchSnapshot();
});

it("should allow cancel", async () => {
  const { getByTestId } = renderWithContextAndRouterForNewItem();

  await waitFor(() => {
    fireEvent.click(getByTestId("cancel"));
  });

  expect(mockHistoryPush.mock.calls.length).toBe(1);
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/2");
  mockHistoryPush.mockClear();
});

it("should require item title", async () => {
  const { getByTestId } = renderWithContextAndRouterForNewItem();

  await waitFor(() => {
    fireEvent.click(getByTestId("submit"));
  });

  expect(getByTestId("titleErrors")).toHaveTextContent("Required");
});

it("should update an existing item", async () => {
  let container;
  await act(async () => {
    container = renderWithContextAndRouterForExistingItem();
  });

  await changeInputValue(container.getByTestId, "title", "updated title");
  await changeInputValue(
    container.getByTestId,
    "description",
    "updated description"
  );

  mockedUpdateListItemUseCase.execute.mockResolvedValue({});

  await waitFor(() => {
    fireEvent.click(container.getByTestId("submit"));
  });

  expect(mockedUpdateListItemUseCase.execute.mock.calls.length).toBe(1);
  const listItem = new ListItem({
    id: 5,
    listId: 2,
    title: "updated title",
    description: "updated description",
  });
  expect(mockedUpdateListItemUseCase.execute.mock.calls[0][0]).toStrictEqual(
    listItem
  );

  expect(mockHistoryPush.mock.calls.length).toBe(1);
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/2");
  mockHistoryPush.mockClear();
});

it("should create a new item", async () => {
  const { getByTestId } = renderWithContextAndRouterForNewItem();

  await changeInputValue(getByTestId, "title", "title");
  await changeInputValue(getByTestId, "description", "description");

  mockedCreateListItemUseCase.execute.mockResolvedValue({});

  await waitFor(() => {
    fireEvent.click(getByTestId("submit"));
  });

  expect(mockedCreateListItemUseCase.execute.mock.calls.length).toBe(1);
  const listItem = new ListItem({
    listId: 2,
    title: "title",
    description: "description",
  });
  expect(mockedCreateListItemUseCase.execute.mock.calls[0][0]).toStrictEqual(
    listItem
  );

  expect(mockHistoryPush.mock.calls.length).toBe(1);
  expect(mockHistoryPush.mock.calls[0][0]).toBe("/lists/2");
  mockHistoryPush.mockClear();
});

const changeInputValue = async (getByTestId, name, value) => {
  await waitFor(() => {
    fireEvent.change(getByTestId(name), {
      target: {
        value: value,
      },
    });
  });
};
