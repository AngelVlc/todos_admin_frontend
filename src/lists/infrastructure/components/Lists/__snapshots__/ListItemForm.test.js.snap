// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot for a new item 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="title"
      >
        Title
      </label>
      <div
        class="control"
      >
        <input
          class="input"
          data-testid="title"
          name="title"
          value=""
        />
      </div>
      <p
        class="help is-danger"
        data-testid="titleErrors"
      />
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="description"
      >
        Description
      </label>
      <div
        class="control"
      >
        <textarea
          class="textarea"
          data-testid="description"
          name="description"
        />
      </div>
      <p
        class="help is-danger"
        data-testid="descriptionErrors"
      />
    </div>
    <div
      class="field is-grouped"
    >
      <div
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          CREATE
        </button>
      </div>
      <div
        class="control"
      >
        <button
          class="button"
          data-testid="cancel"
          type="button"
        >
          CANCEL
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`should match the snapshot for an existing item 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="field"
    >
      <label
        class="label"
        for="title"
      >
        Title
      </label>
      <div
        class="control"
      >
        <input
          class="input"
          data-testid="title"
          name="title"
          value="item title"
        />
      </div>
      <p
        class="help is-danger"
        data-testid="titleErrors"
      />
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="description"
      >
        Description
      </label>
      <div
        class="control"
      >
        <textarea
          class="textarea"
          data-testid="description"
          name="description"
        >
          item description
        </textarea>
      </div>
      <p
        class="help is-danger"
        data-testid="descriptionErrors"
      />
    </div>
    <div
      class="field is-grouped"
    >
      <div
        class="control"
      >
        <button
          class="button"
          data-testid="submit"
          type="submit"
        >
          SAVE
        </button>
      </div>
      <div
        class="control"
      >
        <button
          class="button"
          data-testid="cancel"
          type="button"
        >
          CANCEL
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
